presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/WeightedReservoirSampleStateStrategy.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.differentialentropy;
+
+import com.facebook.presto.operator.aggregation.reservoirsample.WeightedDoubleReservoirSample;
+import com.facebook.presto.spi.PrestoException;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+
+import static com.facebook.presto.operator.aggregation.differentialentropy.EntropyUtils.calculateEntropyFromSamples;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static java.lang.String.format;
+
+public class WeightedReservoirSampleStateStrategy
+        implements DifferentialEntropyStateStrategy
+{
+    private final WeightedDoubleReservoirSample reservoir;
+
+    public WeightedReservoirSampleStateStrategy(long size)
+    {
+        if (size <= 0) {
+            throw new PrestoException(
+                    INVALID_FUNCTION_ARGUMENT,
+                    format("Sample size %s must be positive", size));
+        }
+
+        reservoir = new WeightedDoubleReservoirSample((int) size);
+    }
+
+    private WeightedReservoirSampleStateStrategy(WeightedReservoirSampleStateStrategy other)
+    {
+        reservoir = other.reservoir.clone();
+    }
+
+    private WeightedReservoirSampleStateStrategy(WeightedDoubleReservoirSample reservoir)
+    {
+        this.reservoir = reservoir;
+    }
+
+    @Override
+    public void validateParameters(
+            long bucketCount,
+            double sample,
+            double weight,
+            double min,
+            double max)
+    {
+        throw new PrestoException(
+                INVALID_FUNCTION_ARGUMENT,
+                format("Unsupported parameters for %s", getClass().getSimpleName()));
+    }
+
+    @Override
+    public void validateParameters(
+            long size,
+            double sample,
+            double weight)
+    {
+        if (weight < 0.0) {
+            throw new PrestoException(
+                    INVALID_FUNCTION_ARGUMENT,
+                    format("Weight %s must be non-negative", weight));
```

Comment:
- put short statements on a single line
- move runtime values to the end of the error message to allow for easy grepping when debugging
- add prefix to error message to provide context

```
    @Override
    public void validateParameters(long size, double sample, double weight)
    {
        if (weight < 0.0) {
            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Invalid input to differential_entropy: weight must be not be negative: %s", weight));
        }

        if (size != reservoir.getMaxSamples()) {
            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Invalid input to differential_entropy: inconsistent max samples: %s vs. %s", size, reservoir.getMaxSamples()));
        }
    }
```